From 85ff8d15ce05a00920e8861e9d046eacd7cefb73 Mon Sep 17 00:00:00 2001
From: Karsten Blees <blees@dcon.de>
Date: Thu, 21 Aug 2014 15:43:55 +0200
Subject: [PATCH] de-X-ify lndir.c

Signed-off-by: Karsten Blees <blees@dcon.de>
---
 lndir.c | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/lndir.c b/lndir.c
index 6a8d950..7dc7e46 100644
--- a/lndir.c
+++ b/lndir.c
@@ -49,8 +49,6 @@ in this Software without prior written authorization from The Open Group.
 #include "config.h"
 #endif
 
-#include <X11/Xos.h>
-#include <X11/Xfuncproto.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/stat.h>
@@ -72,7 +70,7 @@ int with_revinfo = 0;		/* -withrevinfo */
 char *rcurdir;
 char *curdir;
 
-static void _X_ATTRIBUTE_PRINTF(2,3) _X_NORETURN
+static void
 quit (int code, const char * fmt, ...)
 {
     va_list args;
@@ -83,14 +81,14 @@ quit (int code, const char * fmt, ...)
     exit (code);
 }
 
-static void _X_NORETURN
+static void
 quiterr (int code, const char *s)
 {
     perror (s);
     exit (code);
 }
 
-static void _X_ATTRIBUTE_PRINTF(1,2)
+static void
 msg (const char * fmt, ...)
 {
     va_list args;
-- 
2.1.0.msysgit.0

