From 5ee3f14b55c085608fae813ac57c0b9c911610d7 Mon Sep 17 00:00:00 2001
From: Karsten Blees <blees@dcon.de>
Date: Thu, 21 Aug 2014 15:43:55 +0200
Subject: [PATCH 2/3] Remove X11 specific code

Signed-off-by: Karsten Blees <blees@dcon.de>
Signed-off-by: Sebastian Schuberth <sschuberth@gmail.com>
---
 lndir.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/lndir.c b/lndir.c
index 22fedd7..372834d 100644
--- a/lndir.c
+++ b/lndir.c
@@ -49,8 +49,6 @@ in this Software without prior written authorization from The Open Group.
 #include "config.h"
 #endif
 
-#include <X11/Xos.h>
-#include <X11/Xfuncproto.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/stat.h>
@@ -72,7 +70,7 @@ int with_revinfo = 0;		/* -withrevinfo */
 char *rcurdir;
 char *curdir;
 
-static void _X_NORETURN
+static void
 help (FILE* f, char * prog_name, int code)
 {
     fprintf(f, "lndir\n");
@@ -93,7 +91,7 @@ help (FILE* f, char * prog_name, int code)
     exit (code);
 }
 
-static void _X_ATTRIBUTE_PRINTF(2,3) _X_NORETURN
+static void
 quit (int code, const char * fmt, ...)
 {
     va_list args;
@@ -104,14 +102,14 @@ quit (int code, const char * fmt, ...)
     exit (code);
 }
 
-static void _X_NORETURN
+static void
 quiterr (int code, const char *s)
 {
     perror (s);
     exit (code);
 }
 
-static void _X_ATTRIBUTE_PRINTF(1,2)
+static void
 msg (const char * fmt, ...)
 {
     va_list args;
-- 
1.9.4.msysgit.1

